udfs/tests/udf_test.py

```diff
@@ -0,0 +1,95 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from os import listdir
+from os import path
+import re
+import unittest
+
+from google.api_core.exceptions import GoogleAPICallError
+from google.cloud.bigquery import QueryJobConfig
+from google.cloud.bigquery.table import _EmptyRowIterator
+from yaml import SafeLoader
+from yaml import load
+
+
+class UDFTest(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        # Client object is initialized in child class
+        cls._client = None
+        cls._test_cases = None
+
+    def _load_test_cases(self, udf_dir):
+        yaml_test_data_path = f'{udf_dir}/test_cases.yaml'
+        if path.isfile(yaml_test_data_path):
+            with open(yaml_test_data_path, 'r') as yaml_file:
+                self._test_cases = load(yaml_file, Loader=SafeLoader)
+
+    def _test_create_udf(self, udf_path, dataset):
+        job_config = QueryJobConfig()
+        job_config.default_dataset = (
+            f'{self._client.project}.{dataset}'
+        )
+        with open(udf_path) as udf_file:
+            try:
+                udf_creation_result = self._client.query(
+                    udf_file.read(),
+                    job_config=job_config
+                ).result()
+                self.assertIsInstance(udf_creation_result, _EmptyRowIterator)
+            except GoogleAPICallError as e:
+                self.fail(e.message)
+
+    def _test_run_udf_and_verify_expected_result(self, udf_path, dataset):
+        udf_name = self.extract_udf_name(udf_path)
+        if self._test_cases.get(udf_name):
+            for case in self._test_cases[udf_name]:
+                try:
+                    actual_result_rows = self._client.query(
+                        f'SELECT {dataset}.{udf_name}('
+                        f' {case["input"]} )'
+                    ).result()
+
+                    expected_result_rows = self._client.query(
+                        f'SELECT {case["expected_output"]}'
+                    ).result()
+
+                    for actual, expected in zip(
+                            actual_result_rows, expected_result_rows):
+                        self.assertEqual(expected, actual)
+                except GoogleAPICallError as e:
+                    self.fail(e.message)
+        else:
+            self.skipTest('Test inputs and outputs are not available')
+
+    @staticmethod
+    def extract_udf_name(udf_path):
+        with open(udf_path) as udf_file:
+            udf_sql = udf_file.read()
+        pattern = re.compile(r'FUNCTION\s*`?(\w+)`?\s*\(')
+        match = pattern.search(udf_sql)
+        if match:
+            udf_name = match[1]
+            return udf_name
+
+    @staticmethod
+    def get_udf_paths(udf_dir_path):
```

Comment:
Suggested redesign:
I think it would be more flexible to specify a base udf path and [walk](https://docs.python.org/3/library/os.html#os.walk) the directory structure looking for sql files.
This would come with the following benefits:
1. Give flexibility for the repo structure to evolve. Imagine if we grow to 100s of community UDFs want to organize community UDFs by directories per topic geo, math, nlp, etc.). 
1. DRY up redundant child test classes: If you look at your child classes (e.g. `TestRedshiftUDF`) they are each 47 lines of code redundant boilerplate that don't do anything but walk the directory structure (by changing `UDF_DIR` in line 22). Instead you could just have a single test class that walks everything in `/udfs`. This way when a new migration candidate (e.g. Hive) is added to the repo, the developer can just author the UDFs and the test yaml spec in the appropriate directory and doesn't have to write a python Test Class because it'd be picked up in the walking process.  

Potential Cons to this approach:
- The tests would all use same BQ client which _might_ slow things down
- You'd have to get clever about the datasets in which the test framework creates these UDFs (I notice you abbreviate teradata-> td_test, redshift -> rs_test). Not sure if that separation is really valuable for integration testing. Could you at least have all of them named consistently with the parent directories they came from rather than abbreviated? Thoughts? 
- This might be seen as over-engineering for the current structure 

In my mind these cons can be overlooked by simplifying the developer / contributing process.

My inspiration for this thought is [how airflow looks for dag files](https://github.com/apache/airflow/blob/63d23ed1dae47ee0bbb42993ef1309d32e5b0c37/airflow/utils/file.py#L79) which is a little complicated as it implements logic similar to .gitignore .